### PR TITLE
Added info about implicit conversions and static_cast 

### PR DESCRIPTION
**Pull request by hazeltonbw**
### Description
Added information about implicit conversions
Added information about static_cast and it's benefits to C++ programmers

### Type of Change

- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/cpp/concepts/data-types/data-types.md [hunk 1]
```diff
@@ -92,14 +92,31 @@ The notation `(type) value` means “convert value to type“. So for example:
...
 
 weight1 = 154.49;
 weight2 = (int) weight1;
+weight3 = weight1; // Implicit conversion
 
 // weight2 is now 154
+// weight3 is also 154 due to the compiler perfoming an implicit conversion
 ```
 
 **Note:** Going from a double to an int simply removes the decimal. There’s no rounding involved.
```
Comment (SSwiniarski) [inline content/cpp/concepts/data-types/data-types.md:105]:
```suggestion
> **Note:** Going from a double to an int simply removes the decimal. There’s no rounding involved.

```

File: content/cpp/concepts/data-types/data-types.md [hunk 2]
```diff
@@ -92,14 +92,31 @@ The notation `(type) value` means “convert value to type“. So for example:
...
+Alternatively, you can use a safer version of casting in C++ called `static_cast`. 
+`static_cast` is performed at compile time, offering better debugging and safer code.
```
Comment (SSwiniarski) [inline content/cpp/concepts/data-types/data-types.md:107]:
```suggestion
`static_cast` is performed at compile time, offering better debugging and safer code.
 
```

File: content/cpp/concepts/data-types/data-types.md [hunk 3]
```diff
@@ -92,14 +92,31 @@ The notation `(type) value` means “convert value to type“. So for example:
...
 **Note:** Going from a double to an int simply removes the decimal. There’s no rounding involved.
+Alternatively, you can use a safer version of casting in C++ called `static_cast`. 
+`static_cast` is performed at compile time, offering better debugging and safer code. 
+```cpp
+double weight1 = 122.59;
+int weight2 = static_cast<int>(weight1);
+std::cout << weight2 << std::endl;
+// 122
+```
+
```
Comment (SSwiniarski) [inline content/cpp/concepts/data-types/data-types.md:114]:
```suggestion

The following example shows how `static_cast` can catch a type conversion error at compile time:

```

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@hazeltonbw, I did a first review and left a couple of suggestions.

**Review by SSwiniarski — APPROVED** [review]
@hazeltonbw, Looks good for review two,.

**Comment by hazeltonbw** [comment]
@SSwiniarski Thank you for reviewing my pull request. This is my first attempt at it, hopefully I'm doing it right. What's the next step in the process?

**Comment by SSwiniarski** [comment]
> @SSwiniarski Thank you for reviewing my pull request. This is my first attempt at it, hopefully I'm doing it right. What's the next step in the process?

@hazeltonbw, generally everything goes through two reviews before it's merged, so one of the other members of the team will come by and review this as well.

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@hazeltonbw Left you some comments/suggestions after reviewing your changes, but had some additional recommendations for other places in the entry that I wanted to mention:

- In the meta-`Description` on line # 3, let's shorten this to (ideally) 150 characters or less.
- Let's rephrase the opening sentence to look like the following (with "data types" in **bold**):
  `C++ supports many **data types** that represent the size and kind of values being stored in memory.`
- Then, let's move the markdown table of data sizes and sizes in the "## Memory Sizes" section up to line # 20 (above the "## Int" section). Then, let's delete the "## Memory Sizes" section.
- On line # 22, replace `-2³¹` to `2³¹` with `-2<sup>31</sup>` and `2<sup>31</sup>`, respectively.
- On line # 40, let's embed "true" and "false" in back-ticks.
- On line # 49, let's embed the single quote mark in back-ticks (i.e., `'`).
- On line # 58, let's link "strings" to here: https://www.codecademy.com/resources/docs/cpp/strings
- On line # 76, replace "your" with "the".

**Review by hazeltonbw — COMMENTED** [review]
Great suggestions, I have included them in my latest commit.

**Review by hazeltonbw — COMMENTED** [review]
Great suggestions, I have included them in my latest commit.

File: content/cpp/concepts/data-types/data-types.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
 Title: 'Data Types'
-Description: 'C++ provides a rather large number of types. However, you can write perfectly good programs using only five of those: int is a type for storing integer (whole) numbers. An integer typically requires 4 bytes of memory space and ranges from -2³¹ to 2³¹. cpp int year = 1991; int age = 28;  double is a type for storing floating point (decimal) numbers. Double variables typically require 8 bytes of memory space.'
+Description: 'C++ supports many **data types** that represent the size and kind of values being stored in memory. You can write perfectly good programs using only five of those.'
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:3]:
```suggestion
Description: 'C++ supports many data type that represent the size and kind of values being stored in memory.'
```

No need for emphasis in the metadata. Also, let's stick with a neutral, manual voice instead of a "you" or "we" voice.

File: content/cpp/concepts/data-types/data-types.md [hunk 5]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
 
 C++ provides a rather large number of types. However, you can write perfectly good programs using only five of those:
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:18]:
```suggestion
C++ supports many **data types** that represent the size and kind of values being stored in memory.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 6]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
...
 
-## Int
+## Memory Size
 
-`int` is a type for storing integer (whole) numbers. An integer typically requires 4 bytes of memory space and ranges from -2³¹ to 2³¹.
+As noted above, the data type of a variable also specifies the size of the information that it can hold:
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:22]:
```suggestion
The size of a given data type is measured in bytes:
```

File: content/cpp/concepts/data-types/data-types.md [hunk 7]
```diff
@@ -94,22 +108,27 @@ double weight1;
...
 
-| Data Type     | Memory Size |
-| ------------- | ----------- |
-| `bool`        | 1 byte      |
-| `char`        | 1 byte      |
-| `int`         | 4 bytes     |
-| `float`       | 4 bytes     |
-| `double`      | 8 bytes     |
-| `std::string` | 24 bytes    |
+**Note:** Not all types can be converted:
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:129]:
```suggestion
> **Note:** Not all types can be converted:
```

File: content/cpp/concepts/data-types/data-types.md [hunk 8]
```diff
@@ -94,22 +108,27 @@ double weight1;
...
+// assigned a double value
 weight2 = (int) weight1;
-
-// weight2 is now 154
+// weight2 is now 154 due to explicit type conversion
+int weight3 = weight1;
+// weight3 is also 154 due to implicit conversion by the compiler
 ```
 
 **Note:** Going from a double to an int simply removes the decimal. There’s no rounding involved.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:118]:
```suggestion
> **Note:** Going from a double to an int simply removes the decimal. There’s no rounding involved.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 9]
```diff
@@ -94,22 +108,27 @@ double weight1;
...
-| `int`         | 4 bytes     |
-| `float`       | 4 bytes     |
-| `double`      | 8 bytes     |
-| `std::string` | 24 bytes    |
+**Note:** Not all types can be converted:
+```cpp
+std::string s = static_cast<std::string>(weight2);
+```
+
+> note: no known conversion for argument 1 from ‘int’ to ‘std::__cxx11::basic_string<char>&&’
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:134]: Is this supposed to be a `> note:` or an error message? If the latter, can we made an edit like this:

```suggestion
This throws the following error:

```

And then attach a \```shell snippet with the error:

```shell
no known conversion for argument 1 from ‘int’ to ‘std::__cxx11::basic_string<char>&&’
````

File: content/cpp/concepts/data-types/data-types.md [hunk 10]
```diff
@@ -94,22 +108,27 @@ double weight1;
...
 
-## Memory Size
+Alternatively, there is a safer version of casting in C++ called `static_cast` that can be used. Performed at compile time, `static_cast` offers better debugging and safer code:
 
-As noted above, the data type of a variable also specifies the size of the information that it can hold:
+```cpp
+double weight1 = 122.03;
+int weight2 = static_cast<int>(weight1);
+std::cout << weight2 << std::endl;
+// 122
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:126]:
```suggestion
// Output: 122
```

File: content/cpp/concepts/data-types/data-types.md [hunk 11]
```diff
@@ -71,17 +84,18 @@ As the name implies, datatype modifiers are used with built-in data types to mod
...
+`const` (constant) variables cannot be changed by the program during execution.
 
 ```cpp
 const double quarter = 0.25;
 
 // and now variable quarter can only be 0.25
 ```
 
-Simply add the keyword `const` before the data type during declaration to make the variable not modifiable.
+Simply add the keyword `const` before the data type to make the variable immutable.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:97]:
```suggestion
```

File: content/cpp/concepts/data-types/data-types.md [hunk 12]
```diff
@@ -71,17 +84,18 @@ As the name implies, datatype modifiers are used with built-in data types to mod
 - `short`
 - `long`
 
-## Const
+## const
 
-`const` (constant) variables cannot be changed by your program during execution.
+`const` (constant) variables cannot be changed by the program during execution.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:89]:
```suggestion
Constant variables cannot be changed by the program during execution. The `const` keyword can be added before the data type to make the variable immutable:
```

File: content/cpp/concepts/data-types/data-types.md [hunk 13]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
...
+| Data Type     | Memory Size |
+| ------------- | ----------- |
+| `bool`        | 1 byte      |
+| `char`        | 1 byte      |
+| `int`         | 4 bytes     |
+| `float`       | 4 bytes     |
+| `double`      | 8 bytes     |
+| `std::string` | 24 bytes    |
+
+## int
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:33]:
```suggestion
## Integers
```

For the most part, we prefer the headers be in "Title Case." What about a pluralization of the data type for a section? (e.g., "Integers")

File: content/cpp/concepts/data-types/data-types.md [hunk 14]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
...
+
+`int` is a type for storing integer (whole) numbers. An integer typically requires 4 bytes of memory space and ranges from -2<sup>31</sup> to 2<sup>31</sup>.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:35]:
```suggestion
`int` is a type for storing integer (whole) numbers. An integer usually requires 4 bytes of memory space and ranges from -2<sup>31</sup> to 2<sup>31</sup>.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 15]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
...
 
 ```cpp
 int year = 1991;
 int age = 28;
 ```
 
-## Double
+## double
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:42]:
```suggestion
## Doubles
```

File: content/cpp/concepts/data-types/data-types.md [hunk 16]
```diff
@@ -35,7 +48,7 @@ double price = 8.99;
 double pi = 3.14159;
 ```
 
-## Bool
+## bool
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:51]:
```suggestion
## Booleans
```

File: content/cpp/concepts/data-types/data-types.md [hunk 17]
```diff
@@ -35,7 +48,7 @@ double price = 8.99;
...
 
 `bool` is a type for storing true or false boolean values. Booleans typically require 1 byte of memory space.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:53]:
```suggestion
The `bool` type stores boolean values of `true` or `false`. These values usually require 1 byte of memory space.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 18]
```diff
@@ -53,9 +66,9 @@ char grade = 'A';
 char punctuation = '?';
 ```
 
-## String
+## string
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:69]:
```suggestion
## Strings
```

File: content/cpp/concepts/data-types/data-types.md [hunk 19]
```diff
@@ -71,17 +84,18 @@ As the name implies, datatype modifiers are used with built-in data types to mod
 - `short`
 - `long`
 
-## Const
+## const
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:87]:
```suggestion
## Constants
```

File: content/cpp/concepts/data-types/data-types.md [hunk 20]
```diff
@@ -17,16 +17,29 @@ CatalogContent:
...
 
 ```cpp
 int year = 1991;
 int age = 28;
 ```
 
-## Double
+## double
 
 `double` is a type for storing floating point (decimal) numbers. Double variables typically require 8 bytes of memory space.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:44]:
```suggestion
The `double` type stores floating point (decimal) numbers. These variables usually require 8 bytes of memory space.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 21]
```diff
@@ -44,7 +57,7 @@ bool organ_donor = true;
 bool late_to_work = false;
 ```
 
-## Char
+## char
 
 `char` is a type for storing individual characters. Characters are wrapped in single quotes '. Characters typically require 1 byte of memory space and range from -128 to 127.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:62]:
```suggestion
The `char` type stores individual characters, wrapped in single quotes `'`. Characters usually require 1 byte of memory space and range from -128 to 127.
```

File: content/cpp/concepts/data-types/data-types.md [hunk 22]
```diff
@@ -44,7 +57,7 @@ bool organ_donor = true;
 bool late_to_work = false;
 ```
 
-## Char
+## char
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:60]:
```suggestion
## Characters
```

File: content/cpp/concepts/data-types/data-types.md [hunk 23]
```diff
@@ -53,9 +66,9 @@ char grade = 'A';
 char punctuation = '?';
 ```
 
-## String
+## string
 
-`std::string` is a type for storing text strings. Strings are wrapped in double quotes `"`.
+[std::string](https://www.codecademy.com/resources/docs/cpp/strings) is a type for storing text strings. Strings are wrapped in double quotes: `"Hello World!"`.
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:71]:
```suggestion
[Strings](https://www.codecademy.com/resources/docs/cpp/strings) are character sequences wrapped in double quotes (e.g., `"Hello World!"`).
The `std::string` type is used for storing text strings. 
```

File: content/cpp/concepts/data-types/data-types.md [hunk 24]
```diff
@@ -53,9 +66,9 @@ char grade = 'A';
...
 
 ```cpp
 std::string message = "good nite";
```
Comment (Dusch4593) [inline content/cpp/concepts/data-types/data-types.md:74]: On the next line, # 75, let's change `@sonnynomnom` to `@codeninja`.

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@hazeltonbw we are _super close_ to getting this merged! 🤩 This last batch of comments/suggestions are really really small. After the next set of changes this PR should be good for merge 😄

**Comment by hazeltonbw** [comment]
> 

Looks good! I like the changes you made, makes everything much clearer.

**Review by Dusch4593 — APPROVED** [review]
@hazeltonbw This looks good! 🚀 Made some minor formatting fixes in be1a943775470a93a60d99ad06bec33a6a251c15 and 4778ac5b07d79d822ccbc1b51db86b46e797488c. Thank you again so much for these edits! 😄 I'll let you know when your contribution goes live.

**Comment by Dusch4593** [comment]
Congrats @hazeltonbw your changes are now live! 🎊 🎉 🥳

Live entry here: https://www.codecademy.com/resources/docs/cpp/data-types

I noticed, however, that your Codecademy profile image wasn't appearing under the list of credited contributors:

Screen Shot 2022-10-01 at 8 26 59 AM

Did you wish for your profile to stay anonymous?

**Comment by hazeltonbw** [comment]
> 

Sweet! If possible I'd like to be under contributions. I linked my GitHub profile to CodeCademy already.

**Comment by Dusch4593** [comment]
If there are edits, I think there might be a slight lag between when the actual content updates and when the contributor profile updates. I believe that this is you, but  I'm wondering why a different profile image? 🤔

Screen Shot 2022-10-03 at 11 45 13 AM

**Comment by hazeltonbw** [comment]
> 

That's me! I guess I never updated my profile picture. It's updated now though! 😄
